system/jlib/jsocket.cpp

```diff
@@ -3435,14 +3511,114 @@ static bool lookupHostAddress(const char *name,unsigned *netaddr)
+        }
+        if (entry->h_addr_list[0]) {
+            unsigned ptr = 0;
+            if (!PreferredSubnet.isNull()) {
+                for (;;) {
+                    ptr++;
+                    if (entry->h_addr_list[ptr]==NULL) {
+                        ptr = 0;
+                        break;
+                    }
+                    IpAddress ip;
+                    ip.setNetAddress(sizeof(unsigned),entry->h_addr_list[ptr]);
+                    if (PreferredSubnet.test(ip))
+                        break;
+                }
+            }
+            memcpy(&netaddr[3], entry->h_addr_list[ptr], sizeof(netaddr[3]));
+            netaddr[2] = 0xffff0000;
+            netaddr[1] = 0;
+            netaddr[0] = 0;
+            return true;
+        }
+        return false;
+    }
+
+#if defined(__linux__) || defined (__APPLE__) || defined(getaddrinfo)
+    bool ret = false;
+    if ( (timeoutms != INFINITE) && (useDNSTimeout()) ) // could addrInfoPool be NULL ?
+    {
+        // getaddrinfo_a() offers an async getaddrinfo method, but has some limitations and a possible mem leak
+        // could implement timeout getaddrinfo functionality without threads by connecting with DNS servers and parsing response ...
+        // lookup performance may be significantly improved if system uses a DNS resolver cache
 
+        PooledThreadHandle thrdHandle;
+        Owned<CAddrInfoThreadArgs> addrThreadInfo = new CAddrInfoThreadArgs(name);
+        CTimeMon dnstimeout(timeoutms);
 
-bool IpAddress::ipset(const char *text)
+        try
+        {
+            // wait up to timeoutms for an available thread from pool ...
+            thrdHandle = addrInfoPool->start((void *)addrThreadInfo.getLink(), "getaddrinfo-thread", timeoutms);
+            // MCK TODO: if have to create a new thread and at os/system/shell thread limit, create/start can take up to several seconds, can we control this ?
+        }
+        catch (IException *e)
+        {
+            addrThreadInfo->Release();
+            StringBuffer excMsg;
+            StringBuffer emsg;
+            emsg.appendf("getaddrinfo failed (thread) (%d) (exc:%d) %s", timeoutms, e->errorCode(), e->errorMessage(excMsg).str());
+            RecursionSafeLogErr(100, 1, emsg.str(), __LINE__, name);
+            e->Release();
+            return false;
+        }
+        catch (...)
+        {
+            addrThreadInfo->Release();
+            RecursionSafeLogErr(101, 1, "getaddrinfo failed (thread) (other exc)", __LINE__, name);
+            return false;
+        }
+
+        // take into account time already passed creating/starting thread above ...
+        unsigned remaining;
+        dnstimeout.timedout(&remaining);
+        if (addrInfoPool->join(thrdHandle, remaining))
+        {
+            if (addrThreadInfo->result)
+            {
+                memcpy(netaddr, addrThreadInfo->netaddr, sizeof(addrThreadInfo->netaddr));
+                return true;
+            }
+        }
+
```

Comment:
Question:  What happens if it fails to join?  Will the thread in the thread poll be retained until join is called()?
Answer: It is ok - the thread is not kept alive in the same way as a normal thread.  Might be worth having a comment.

Reply:
I've added a comment